File: ndb-bindings-0.7.1/perl/async.pl
#     ndb-bindings: Bindings for the NDB API
#       Copyright (C) 2006 MySQL, Inc.
    
#       This program is free software; you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation; either version 2 of the License, or 
#     (at your option) any later version.
    
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.

#     You should have received a copy of the GNU General Public License
#     along with this program; if not, write to the Free Software
#     Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


use lib 'blib/arch';
use lib 'blib';
use lib 'blib/lib';
use ndbapi;
use DBI;
use POSIX qw( ceil floor );
use Time::HiRes qw( gettimeofday tv_interval sleep );

$num_iter = shift;
$INSERT_NUM = shift;
do { print "Usage:\n\ttest.pl NUM_OF_ITERATIONS NUM_OF_ROWS\n";
  exit -1; }  unless $INSERT_NUM;
#$BATCH_SIZE = 100;
$BATCH_SIZE = 1;

### Connect to cluster

print "Dropping and recreating scehma\n\n";

$dbh = DBI->connect("dbi:mysql:host=localhost;read_default_group=client","root","" );


$dbh->do("CREATE DATABASE if not exists test");
$dbh->do("USE test");

$dbh->do("DROP TABLE if exists mytablename");
$dbh->do("CREATE TABLE
  mytablename
   (ATTR1 INT UNSIGNED not null,
    ATTR2 INT UNSIGNED NOT NULL,
    PRIMARY KEY USING HASH (ATTR1)
   )
ENGINE=NDBCLUSTER
");

print "connecting to cluster\n";

$connection = new ndbapi::NdbClusterConnection();


if ($connection->connect(5,3,1)) {

    print "Connect to cluster management server failed.\n";
    exit -1;
}


if ($connection->wait_until_ready(30,30)) {
    print "Cluster was not ready within 30 secs.\n";
    exit -1;
} else { print "cluster ready\n"; }

$myNdb = $connection->createNdb("test");


if ($myNdb->init(4) == -1) {
    print $myNdb->getNdbError()->getMessage();

    exit -1;
}

print "running tests: \n";


my $before_time = [gettimeofday];

my $auto_id=0;

my $done=0;
foreach my $t (1..(ceil($INSERT_NUM/$BATCH_SIZE))) {

    my $myTransaction = $myNdb->startTransaction();
    while (! $myTransaction ) {
	sleep 1/1000;
	$myNdb->sendPollNdb(0);
    	$myTransaction = $myNdb->startTransaction();
    }

    my $val = $t*$BATCH_SIZE-$INSERT_NUM;
    my $offset = ($val > 0 ) ? $val : 0;

    foreach my $i (1..($BATCH_SIZE-$offset)) {

      my $myOperation = $myTransaction->getNdbOperation("mytablename");
      $myOperation->insertTuple();

      #my $auto_id = $myNdb->getAutoIncrementValue("mytablename",$BATCH_SIZE);
      $auto_id++;
      #print "$t - $i - $auto_id - $myOperation\n";

      $myOperation->equal("ATTR1", $auto_id);
      $myOperation->setVal("ATTR2",$i);

    }

    #my $ret = $myTransaction->execute($ndb::Commit);
    my $ret;
    {
	my $trans=$myTransaction;
    	$ret = $myTransaction->executeAsynchPrepare($ndb::Commit,
		sub { $done--; 
		print "ASYNC CALLED\n";
    		# We want to close transactions after the callback
    	         } );
    }

    if ($ret == -1) {
      print $myTransaction->getNdbError()->getMessage(), "\n";
    }
    $done++;
}

while ( $done>0 ) {
	$myNdb->sendPollNdb(3000,1);
	sleep 1/1000;
}

my $elapsed = tv_interval ( $before_time );

print "Insert time for $INSERT_NUM:\t$elapsed\n";


## Fetch list of ids. We could do this when we build it
## But then we couldn't test Scan Operations

my @ids;

my $myTransaction = $myNdb->startTransaction();


my $myScanOperation=$myTransaction->getNdbScanOperation("mytablename");

print $myTransaction->getNdbError()->getMessage() unless $myScanOperation;

if ($myScanOperation->readTuples($ndb::NdbOperation::LM_CommittedRead)) {
  print $myScanOperation->getNdbError()->getMessage();
}



my $myRecAttr=$myScanOperation->getValue("ATTR1");

$myTransaction->execute($ndb::NoCommit);

while ($myScanOperation->nextResultMt() == 0) {
  
  my $random_id = $myRecAttr->u_32_value();
  push(@ids,$random_id);
  
}

$myTransaction->close();


#Fetch records using NDBAPI

my $before_time = [gettimeofday];

foreach my $f (0..$num_iter) {

      my $myTransaction = $myNdb->startTransaction();
      if (!$myTransaction) {
        print $myNdb->getNdbError()->getMessage(); }

      my $myOper = $myTransaction->getNdbOperation("mytablename");
      $myOper->readTuple($NdbOperation::LockMode::LM_Read);

      my $id_num = $ids[rand($INSERT_NUM)];

      $myOper->equal("ATTR1",$id_num);

      my $myRecAttr = $myOper->getValue("ATTR2");

      if ($myTransaction->execute( $ndb::Commit ) == -1) {
        print $myTransaction->getNdbError()->getMessage(),"\n"; 
      }
      $foo=$myRecAttr->u_32_value();

      $myTransaction->close();

    }


my $elapsed = tv_interval ( $before_time );
print "Fetch time for NDBAPI - $num_iter:\t$elapsed\n";

File: ndb-bindings-0.7.1/python/examples/async.py
from __future__ import with_statement
import sys,time,random,struct,math
from mysql.cluster import ndbapi
import mysql.cluster.ndbapi
import threading 

class PythonCallback(ndbapi.BaseCallback):

  def __init__(self, recAttr):
    self.recAttr=recAttr

  def __call__(self, *args, **kw): 
    self.callback(*args,**kw)

  def callback(self, ret, myTrans):
    #print "value = ", self.recAttr.getValue()
    x=self.recAttr.getValue()

import  MySQLdb

if len(sys.argv) != 3:
  print "Usage:\n\ttest.py NUM_OF_ITERATIONS NUM_OF_ROWS "
  sys.exit(1)

num_iter=int(sys.argv[1])
INSERT_NUM=int(sys.argv[2])
concurrent=1
BATCH_SIZE=1000

db = MySQLdb.connect(host="localhost",user="root",read_default_group="client")

cur=db.cursor()

print "Dropping and recreating schema\n"

cur.execute("CREATE DATABASE if not exists test")
cur.execute("USE test")

cur.execute("drop table if exists mytablename")

cur.execute("""
CREATE TABLE if not exists
  mytablename
   (ATTR1 INT UNSIGNED not null auto_increment,
    ATTR2 INT UNSIGNED NOT NULL,
    PRIMARY KEY USING HASH (ATTR1))
ENGINE=NDBCLUSTER 
""")


### Connect to cluster

print "connecting to cluster\n"
   
try:
  connection = ndbapi.NdbClusterConnection();

  connection.connect(1,1,1)
except mysql.cluster.ndbapi.NdbApiException,e:
  print e
  sys.exit(-1)

if (connection.waitUntilReady(30,30)):
    print "Cluster was not ready within 30 secs."
    sys.exit(-1);


myNdb = connection.createNdb("test")

if (myNdb.init(1000) == -1):
    print myNdb.getNdbError().getMessage()
    sys.exit(-1)


print "running tests:\n"

### Fill db


before_t = time.time()

for t in range(0,int(math.ceil((int(INSERT_NUM)*1.0)/(1.0*BATCH_SIZE)))):
  
  myTransaction= myNdb.startTransaction()
  val = ((t+1)*BATCH_SIZE)-INSERT_NUM
  offset = 0
  if ( val > 0 ):
    offset = val

  for i in range(0,BATCH_SIZE-offset):
    
    myOperation=myTransaction.getNdbOperation("mytablename")
    myOperation.insertTuple()
    auto_id = myNdb.getAutoIncrementValue("mytablename",BATCH_SIZE)
   
    myOperation.equalUint64("ATTR1",auto_id);
    myOperation.setValue("ATTR2", i);
      
  ret = myTransaction.execute( ndbapi.Commit )
  if ret == -1:
    print myTransaction.getNdbError().getMessage()
  
  #myTransaction.close()

after_t = time.time()
print "Insert time for %s: %s -- %s" % ( INSERT_NUM, after_t, before_t)
print "\t %s" % (after_t - before_t)


### Get list of ids
myTransaction = myNdb.startTransaction()

myScanOperation=myTransaction.getNdbScanOperation("mytablename");


if myScanOperation is None:
    print myTransaction.getNdbError().getMessage()


if myScanOperation.readTuples(ndbapi.NdbOperation.LM_CommittedRead):
    print myScanOperation.getNdbError().getMessage()


    
myRecAttr=myScanOperation.getValue("ATTR1");
    
    
myTransaction.execute(ndbapi.NoCommit);

ids=[]
while 1:
      
    if (myScanOperation.nextResult(True) != 0) :
        break
      
    random_id = myRecAttr.getValue()
    ids.append(random_id)
      
    
#myTransaction.close()


### Fetch ids using ndb

before_t = time.time()

threadlist=[]
connsem=threading.BoundedSemaphore(value=1)

def runit():
  mydata=threading.local()
  mydata.t=threading.currentThread()
  with connsem:
    mydata.ndb=connection.createNdb("test")
  if (mydata.ndb.init(4) == -1):
    print mydata.ndb.getNdbError().getMessage()
    sys.exit(-1)
  mydata.translist=[]

  for mydata.f in range(0,int(num_iter/concurrent)):

    mydata.id_num = ids[random.randrange(0,len(ids))]
    #print "starting new trans", mydata.f,mydata.t.getName()
    mydata.myTrans = mydata.ndb.startTransaction("mytablename","%s"%mydata.id_num)
    
    if mydata.myTrans is None:
        print mydata.ndb.getNdbError().getMessage()

    mydata.myOper = mydata.myTrans.getNdbOperation("mytablename")
    mydata.myOper.readTuple(ndbapi.NdbOperation.LM_Read)

    mydata.myOper.equalUint64("ATTR1",mydata.id_num)
            
    mydata.myRecAttr= mydata.myOper.getValue("ATTR2")

    mydata.cb =  PythonCallback(mydata.myRecAttr)
    
    if mydata.myTrans.executeAsynchPrepare( ndbapi.Commit , mydata.cb ) == -1:
        print mydata.myTrans.getNdbError().getMessage()
    mydata.translist.append(mydata.myTrans)
    if mydata.f%10==0:
      #print "polling"
      mydata.ndb.sendPollNdb(3000,20)
      for mydata.x in range(0,len(mydata.translist)):
        mydata.x=mydata.translist.pop()
        mydata.x.close()
        del(mydata.x)
  time.sleep(0.001)
  mydata.ndb.sendPollNdb(3000,4)

  for mydata.x in range(0,len(mydata.translist)):
    mydata.x=mydata.translist.pop()
    mydata.x.close()
    del(mydata.x)

for x in range(0,concurrent):
  thr=threading.Thread(target=runit,name="thread%s"%x)
  threadlist.append(thr)
  thr.start()

for x in threadlist:
  x.join()

after_t = time.time()
print "NDBAPI time for %s: %s -- %s" % ( num_iter, after_t, before_t)
print "\t %s" % (after_t - before_t)



cur=db.cursor()

before_t = time.time()

for f in range(0,int(num_iter)):

    id_num = ids[random.randrange(0,len(ids))]

    cur.execute("select ATTR2 from mytablename where ATTR1=%s" % (id_num))
    res=cur.fetchall()

after_t = time.time()

after_t = time.time()
print "MySQL time for %s: %s -- %s" % ( num_iter, after_t, before_t)
print "\t %s" % (after_t - before_t)

#ndbapi.ndb_end()
